### PR TITLE
Fix ServiceWorker redirects

### LINKED ISSUE #593: ServiceWorker and protected routes
**Describe the bug**
SW is unable to handle redirects for protected routes on the client-side.

**To Reproduce**
1. Go to https://demo.deity.io/account/orders/1
2. See the error

```
The FetchEvent for "https://demo.deity.io/account/orders/1" resulted in a network error response: a redirected response was used for a request whose redirect mode is not "follow".
```

### PR DESCRIPTION
**Pull request by nimobeeren**
### Please check if the PR fulfills these requirements ###

- [x] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Changelog have been added / updated (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
Bug fix.

**What is the current behavior? (You can also link to an open issue here)**
Fixes #593.

**What is the new behavior (if this is a feature change)?**
Although the redirect works, a new error is printed to console:

`index.js:1 Warning: Prop `href` did not match. Server: "/sign-in" Client: "/account/orders/1"`

Not sure how to fix that.

**Does this PR introduce a breaking change? (What changes might users need to make in their application due to this PR?)**
No.

**Other information:**
Based on https://stackoverflow.com/questions/45434470/only-in-chrome-service-worker-a-redirected-response-was-used-for-a-reque.

### CODE REVIEW

File: packages/falcon-client/src/serviceWorker/sw.js [hunk 1]
```diff
@@ -39,10 +39,26 @@ if (CONFIG.precache) {
...
+  });
+}
+
 const router = new Router();
 self.addEventListener('fetch', event => {
-  const responsePromise = router.handleRequest(event);
+  let responsePromise = router.handleRequest(event);
   if (responsePromise) {
+    responsePromise = responsePromise.then(res => (res.redirected ? cleanResponse(res) : res));
     event.respondWith(responsePromise);
```
Comment (tutok) [inline packages/falcon-client/src/serviceWorker/sw.js]: can not we do 
```suggestion
     event.respondWith(responsePromise.then(res => (res.redirected ? cleanResponse(res) : res)));
```
?
then `responsePromise` can be still const.

The second thing is why we need to do "response to blob rendering" only when there is redirect? should not we do it for all possible responses?

File: packages/falcon-client/src/serviceWorker/sw.js [hunk 2]
```diff
@@ -39,10 +39,26 @@ if (CONFIG.precache) {
   precacheAndRoute(ENTRIES, {});
 }
 
+async function cleanResponse(response) {
```
Comment (tutok) [inline packages/falcon-client/src/serviceWorker/sw.js]: I am not sure if the name of this function describes what is doing

**Comment by Borales** [comment]
@nimobeeren should we consider including "external redirects" into this PR also? 

Like `<Redirect to="http://example.com" />`

_Although, it also could be considered as a separate task._

**Comment by nimobeeren** [comment]
> @nimobeeren should we consider including "external redirects" into this PR also?

@Borales are those also broken then?

**Comment by Borales** [comment]
@nimobeeren looks like so, I was testing such redirection, it was partially working, and when it was working - I got redirected to `http://localhost:3000/http://example.com`

**Comment by nimobeeren** [comment]
@Borales I can take a look at that within this PR. Marking this as WIP

File: packages/falcon-client/src/serviceWorker/sw.js [hunk 3]
```diff
@@ -43,7 +43,9 @@ const router = new Router();
 self.addEventListener('fetch', event => {
   const responsePromise = router.handleRequest(event);
   if (responsePromise) {
-    event.respondWith(responsePromise);
+    // if response is redirected, we should clone the response before sending it
+    // see https://stackoverflow.com/a/45440505/412319
+    event.respondWith(responsePromise.then(res => (res.redirected ? res.clone() : res)));
```
Comment (tutok) [inline packages/falcon-client/src/serviceWorker/sw.js]: @nimobeeren why the previous `cleanResponse` function was removed?

```suggestion
    event.respondWith(responsePromise.then(res => (res.redirected ? cleanResponse(res) : res)));
```
I did a couple of tests, and the current implementation does not work on my side. I am receiving the same error as in reported issue.

Alo did som tests with suggested SO solution and it looks fine. Maybe we misunderstood in one of the previous comments?

Comment (mateuszwozniak) [reply packages/falcon-client/src/serviceWorker/sw.js]: I also did some tests and it looks like @tutok is right - the solution with `cleanResponse(res)` works correctly for me too, while version from this PR doesn't so @nimobeeren please update it.

**Pull request merged**